File: SL_Compiler_Tests/CodeGenerationTests.cpp
#include "stdafx.h"
#include "CppUnitTest.h"
#include <stdio.h>
#include <memory.h>
#include <windows.h>
#include <assert.h>
using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace SL_Compiler_Tests
{
	char data[1<<16];

		// __stdcall makes the callee pop the arguments from the stack upon return
		void __stdcall Put(char ch) 
		{
			printf("%c", ch);
		}
		void __stdcall PutLn() 
		{
			printf("\n");
		}

		void __stdcall Assert(int exp, int act)
		{
			Assert::AreEqual(exp, act);
		}

		#pragma endregion 

		void ExecuteFile(char* filename)
		{
			int i = 0x0FFFFFFF;
			int j = 0x0FFFFFFF;
			int z = i * j;
				/* load the code file */ 
			FILE* fin;
			fopen_s(&fin, filename, "rb");
			if (fin == NULL) 
			{
				printf("couldn't open %s\n", filename);
				Assert::Fail();
			}

			byte* code = (byte*) VirtualAllocEx(GetCurrentProcess(), 0, 1<<16, MEM_COMMIT, PAGE_EXECUTE_READWRITE);
			size_t size = fread(code, 1, 1<<16, fin);
			fclose(fin);

			printf("loaded %i bytes\n", size);
			printf("executing...\n");
	
			/* setup the global data block */
			memset(data, 0, sizeof(data));
			((void**)data)[0] = (void*)Put;
			((void**)data)[1] = (void*)PutLn;
			((void**)data)[2] = (void*)Assert;

			/* set edi and jump to the code block */
			void* start = code;
			#ifdef __GNUC__
			  __asm__ ("call *%%eax" : /* */ : "a" (start), "D" (data) : /* */ );
			#else
				__asm {
					lea edi, data;
 					call start;
				};
				
			
			#endif
		}


	TEST_CLASS(UnitTest1)
	{
	public:
		
		TEST_METHOD(GlobalAssignment_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\GlobalAssignment.obj");
		}
		
		TEST_METHOD(GlobalAssignment2_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\GlobalAssignment2.obj");
		}

		TEST_METHOD(LocalAssignment_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\LocalAssignment.obj");
		}

		TEST_METHOD(SimpleAdd_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\SimpleAdd.obj");
		}

		TEST_METHOD(SimpleMeth_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\SimpleMeth.obj");
		}

		TEST_METHOD(SimpleMeth2_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\SimpleMeth2.obj");
		}

		TEST_METHOD(PutTest_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\PutTest.obj");
		}

		TEST_METHOD(Param1_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\Param1.obj");
		}

		TEST_METHOD(ManyParams_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\ManyParams.obj");
		}

		TEST_METHOD(Locals_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\Locals.obj");
		}

		TEST_METHOD(SimpleExpression1_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\SimpleExpression1.obj");
		}

		TEST_METHOD(SimpleExpression2_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\SimpleExpression2.obj");
		}

		TEST_METHOD(SimpleExpression3_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\SimpleExpression3.obj");
		}

		TEST_METHOD(SimpleExpression4_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\SimpleExpression4.obj");
		}

		TEST_METHOD(Branch1_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\Branch1.obj");
		}

		TEST_METHOD(Branch2_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\Branch2.obj");
		}

		TEST_METHOD(While1_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\While1.obj");
		}

		TEST_METHOD(ORDCHRTest_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\ORDCHRTest.obj");
		}

		TEST_METHOD(BiggerExample_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\BiggerExample.obj");
		}

		TEST_METHOD(Faculty_Code)
		{
			ExecuteFile("C:\\Martin\\Uni\\Übersetzerbau2\\SL-Compiler\\SL_Compiler\\Tests\\CodeGeneration\\Faculty.obj");
		}
	};
}

File: SL_Compiler_Tests_Builder/CodeGenerationTests.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace SL_Compiler_Tests_Builder
{
    [TestClass]
    public class CodeGenerationTests
    {
        [TestMethod]
        public void GlobalAssignment()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\GlobalAssignment.sl");
        }

        [TestMethod]
        public void UndefNameMeth()
        {
            TestSupport.Init();
            TestSupport.ExpectError("-- line 3 col 1: Method not found");
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\UndefNameMeth.sl");
        }

        [TestMethod]
        public void GlobalAssignment2()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\GlobalAssignment2.sl");
        }

        [TestMethod]
        public void LocalAssignment()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\LocalAssignment.sl");
        }

        [TestMethod]
        public void SimpleAdd()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\SimpleAdd.sl");
        }

        [TestMethod]
        public void SimpleMeth()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\SimpleMeth.sl");
        }

        [TestMethod]
        public void SimpleMeth2()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\SimpleMeth2.sl");
        }

        [TestMethod]
        public void PutTest()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\PutTest.sl");
        }

        [TestMethod]
        public void Param1()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\Param1.sl");
        }

        [TestMethod]
        public void ManyParams()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\ManyParams.sl");
        }

        [TestMethod]
        public void Locals()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\Locals.sl");
        }

        [TestMethod]
        public void WrongMethAssign1()
        {
            TestSupport.Init();
            TestSupport.ExpectError("-- line 8 col 9: Left operand is not a variable");
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\WrongMethAssign1.sl");
        }

        [TestMethod]
        public void WrongMethAssign2()
        {
            TestSupport.Init();
            TestSupport.ExpectError("-- line 8 col 10: called object is not a method");
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\WrongMethAssign2.sl");
        }

        [TestMethod]
        public void WrongMethAssign3()
        {
            TestSupport.Init();
            TestSupport.ExpectError("-- line 5 col 10: void method must not return a value");
            TestSupport.ExpectError("-- line 8 col 7: Invalid call of void method");
            TestSupport.ExpectError("-- line 8 col 11: Members are not compatible");
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\WrongMethAssign3.sl");
        }

        [TestMethod]
        public void WrongMethAssign4()
        {
            TestSupport.Init();
            TestSupport.ExpectError("-- line 8 col 11: Members are not compatible");
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\WrongMethAssign4.sl");
        }

        [TestMethod]
        public void WrongReturn()
        {
            TestSupport.Init();
            TestSupport.ExpectError("-- line 5 col 10: return type must match method type");
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\WrongReturn.sl");
        }

        [TestMethod]
        public void NoReturn()
        {
            TestSupport.Init();
            TestSupport.ExpectError("-- line 5 col 3: Return expression required");
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\NoReturn.sl");
        }

        [TestMethod]
        public void NoReturn2()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\NoReturn2.sl");
        }

        [TestMethod]
        public void WrongReturn2()
        {
            TestSupport.Init();
            TestSupport.ExpectError("-- line 5 col 10: void method must not return a value");
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\WrongReturn2.sl");
        }

        [TestMethod]
        public void WrongParamType()
        {
            TestSupport.Init();
            TestSupport.ExpectError("-- line 7 col 6: parameter type mismatch");
            TestSupport.ExpectError("-- line 7 col 9: parameter type mismatch");
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\WrongParamType.sl");
        }

        [TestMethod]
        public void WrongNumberActParameters()
        {
            TestSupport.Init();
            TestSupport.ExpectError("-- line 8 col 11: Number of actual and formal parameters does not match");
            TestSupport.ExpectError("-- line 9 col 17: Number of actual and formal parameters does not match");

            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\WrongNumberActParameters.sl");
        }

        [TestMethod]
        public void SimpleExpression1()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\SimpleExpression1.sl");
        }

        [TestMethod]
        public void SimpleExpression2()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\SimpleExpression2.sl");
        }

        [TestMethod]
        public void SimpleExpression3()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\SimpleExpression3.sl");
        }

        [TestMethod]
        public void SimpleExpression4()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\SimpleExpression4.sl");
        }

        [TestMethod]
        public void Branch1()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\Branch1.sl");
        }

        [TestMethod]
        public void Branch2()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\Branch2.sl");
        }

        [TestMethod]
        public void Branch3()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\Branch3.sl");
        }

        [TestMethod]
        public void While1()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\While1.sl");
        }

        [TestMethod]
        public void ORDCHRTest()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\ORDCHRTest.sl");
        }

        [TestMethod]
        public void BiggerExample()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\BiggerExample.sl");
        }

        [TestMethod]
        public void Faculty()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\Faculty.sl");
        }

        [TestMethod]
        public void Faculty1()
        {
            TestSupport.Init();
            TestSupport.Parse(@"C:\Martin\Uni\Übersetzerbau2\SL-Compiler\SL_Compiler\Tests\CodeGeneration\Faculty1.sl");
        }
    }
}
